### PR TITLE
feat(client): Pass Google Analytics query params to metrics.

### LINKED ISSUE #2579: Consume & report GA tags to metrics?
A possible idea to help out the growth team - consume Google Analytics query parameters and report them to our metrics infrastructure so they can run similar queries everywhere.

cc @rfk, @jpetto, @habber

ref #2399 
ref https://bugzilla.mozilla.org/show_bug.cgi?id=1162192

### PR DESCRIPTION
**Pull request by shane-tomlinson**
* Reliers can specify Google Analytics query parameters to be logged/sent to our metrics infrastructure.
* Consume the following Google Analytics query parameters:
  * utm_campaign
  * utm_content
  * utm_medium
  * utm_source
  * utm_term

fixes #2579
ref #2101

### CODE REVIEW

**Comment by pdehaan** [comment]
@shane-tomlinson Looks like we need to disable some camel-case variable warnings (thanks ESLint)...

Two options:
(1) per line (x30 times):
```js
utm_campaign: relier.get('utm_campaign'), // eslint-disable-line camelcase
```

(2) per block (x6 times):
```js
/*eslint-disable camelcase */
utm_campaign: relier.get('utm_campaign'),
utm_content: relier.get('utm_content'),
utm_medium: relier.get('utm_medium'),
utm_source: relier.get('utm_source'),
utm_term: relier.get('utm_term')
/*eslint-enable camelcase */
```

### Current ESLint results:
```
Running "eslint:app" (eslint) task

app/scripts/lib/app-start.js
  257:8  error  Identifier 'utm_campaign' is not in camel case  camelcase
  258:8  error  Identifier 'utm_content' is not in camel case   camelcase
  259:8  error  Identifier 'utm_medium' is not in camel case    camelcase
  260:8  error  Identifier 'utm_source' is not in camel case    camelcase
  261:8  error  Identifier 'utm_term' is not in camel case      camelcase
app/scripts/lib/metrics.js
   69:2  error  Function 'Metrics' has a complexity of 23       complexity
  188:8  error  Identifier 'utm_campaign' is not in camel case  camelcase
  189:8  error  Identifier 'utm_content' is not in camel case   camelcase
  190:8  error  Identifier 'utm_medium' is not in camel case    camelcase
  191:8  error  Identifier 'utm_source' is not in camel case    camelcase
  192:8  error  Identifier 'utm_term' is not in camel case      camelcase
app/scripts/models/reliers/relier.js
  37:6  error  Identifier 'utm_campaign' is not in camel case  camelcase
  38:6  error  Identifier 'utm_content' is not in camel case   camelcase
  39:6  error  Identifier 'utm_medium' is not in camel case    camelcase
  40:6  error  Identifier 'utm_source' is not in camel case    camelcase
  41:6  error  Identifier 'utm_term' is not in camel case      camelcase
app/scripts/models/resume-token.js
  22:6  error  Identifier 'utm_campaign' is not in camel case  camelcase
  23:6  error  Identifier 'utm_content' is not in camel case   camelcase
  24:6  error  Identifier 'utm_medium' is not in camel case    camelcase
  25:6  error  Identifier 'utm_source' is not in camel case    camelcase
  26:6  error  Identifier 'utm_term' is not in camel case      camelcase
app/tests/spec/lib/metrics.js
  42:8  error  Identifier 'utm_campaign' is not in camel case  camelcase
  43:8  error  Identifier 'utm_content' is not in camel case   camelcase
  44:8  error  Identifier 'utm_medium' is not in camel case    camelcase
  45:8  error  Identifier 'utm_source' is not in camel case    camelcase
  46:8  error  Identifier 'utm_term' is not in camel case      camelcase
app/tests/spec/models/reliers/relier.js
  55:10  error  Identifier 'utm_campaign' is not in camel case  camelcase
  56:10  error  Identifier 'utm_content' is not in camel case   camelcase
  57:10  error  Identifier 'utm_medium' is not in camel case    camelcase
  58:10  error  Identifier 'utm_source' is not in camel case    camelcase
  59:10  error  Identifier 'utm_term' is not in camel case      camelcase

✖ 31 problems (31 errors, 0 warnings)
Warning: Task "eslint:app" failed. Use --force to continue.
Aborted due to warnings.
```

File: app/scripts/lib/metrics.js [hunk 1]
```diff
@@ -64,40 +69,42 @@ define([
   function Metrics (options) {
     /*eslint complexity: [2, 18] */
```
Comment (pdehaan) [inline app/scripts/lib/metrics.js]: According to ESLint, our complexity jumped from `18` to `23` here.

**Workaround:**
```js
/*eslint complexity: [2, 23] */
```


```
app/scripts/lib/metrics.js
   69:2  error  Function 'Metrics' has a complexity of 23       complexity
```

File: app/scripts/models/reliers/relier.js [hunk 2]
```diff
@@ -4,25 +4,41 @@
...
       allowCachedCredentials: true,
+      campaign: null,
+      email: null,
       entrypoint: null,
-      campaign: null
+      preVerifyToken: null,
+      service: null,
+      setting: null,
+      uid: null,
+      utm_campaign: null,
```
Comment (shane-tomlinson) [inline app/scripts/models/reliers/relier.js]: Add these to the [query-params doc](https://github.com/mozilla/fxa-content-server/blob/master/docs/query-params.md).

**Comment by vladikoff** [comment]
Hey @shane-tomlinson is still WIP? What else are we missing?

**Comment by shane-tomlinson** [comment]
ref #2688

File: app/scripts/lib/app-start.js [hunk 3]
```diff
@@ -264,18 +264,23 @@ function (
       var relier = this._relier;
       var screenInfo = new ScreenInfo(this._window);
       this._metrics = this._createMetrics(this._config.metricsSampleRate, {
-        lang: this._config.language,
-        service: relier.get('service'),
-        context: relier.get('context'),
-        entrypoint: relier.get('entrypoint'),
-        migration: relier.get('migration'),
+        able: this._able,
```
Comment (shane-tomlinson) [inline app/scripts/lib/app-start.js]: @vladikoff and @zaach gave me a bit of a verbal assault for putting things in alphabetical order and making their review job more difficult. Back the alphabetical ordering changes out and do it in a separate PR.

File: app/scripts/lib/metrics.js [hunk 4]
```diff
@@ -62,42 +67,44 @@ define([
   }
 
   function Metrics (options) {
-    /*eslint complexity: [2, 18] */
+    /*eslint complexity: [2, 23] */
     options = options || {};
+    var self = this;
 
-    // by default, send the metrics to the content server.
-    this._collector = options.collector || '';
```
Comment (shane-tomlinson) [inline app/scripts/lib/metrics.js]: Same here w/ the alphabetizing.

File: docs/query-params.md [hunk 5]
```diff
@@ -82,6 +82,30 @@ When the relier supports being linked to its `redirect_uri` without extra OAuth
 
 ## Firefox/Sync parameters
 
+### `customizeSync`
```
Comment (shane-tomlinson) [inline docs/query-params.md]: It looks like there are a ton of tangential doc updates. Separate PR.

File: app/scripts/lib/metrics.js [hunk 6]
```diff
@@ -93,6 +98,13 @@ define([
     this._screenHeight = options.screenHeight || NOT_REPORTED_VALUE;
     this._screenWidth = options.screenWidth || NOT_REPORTED_VALUE;
 
+    this._referrer = this._window.document.referrer || NOT_REPORTED_VALUE,
```
Comment (shane-tomlinson) [inline app/scripts/lib/metrics.js]: `referrer` was in our list of allowed metrics but never reported.

File: app/scripts/models/reliers/base.js [hunk 7]
```diff
@@ -83,6 +83,17 @@ define([
       return {};
     },
 
+    pickAs: function (definition) {
```
Comment (shane-tomlinson) [inline app/scripts/models/reliers/base.js]: Not used, delete

File: app/tests/spec/lib/metrics.js [hunk 8]
```diff
@@ -24,21 +24,27 @@ function (chai, $, p, Metrics, AuthErrors, WindowMock, TestHelpers) {
 
     beforeEach(function () {
       windowMock = new WindowMock();
+      windowMock.document.referrer = 'https://marketplace.firefox.com';
 
       metrics = new Metrics({
-        window: windowMock,
```
Comment (shane-tomlinson) [inline app/tests/spec/lib/metrics.js]: de-alphabetize

**Comment by shane-tomlinson** [comment]
@vladikoff - ping!

File: app/tests/spec/lib/metrics.js [hunk 9]
```diff
@@ -74,11 +80,18 @@ function (chai, $, p, Metrics, AuthErrors, WindowMock, TestHelpers) {
         assert.equal(filteredData.migration, 'sync1.5');
         assert.equal(filteredData.campaign, 'fennec');
 
+        assert.equal(filteredData.referrer, 'https://marketplace.firefox.com');
```
Comment (shane-tomlinson) [inline app/tests/spec/lib/metrics.js]: A bit of a tangential addition. `referrer` was on the list of allowed fields, but did not have a test.

**Comment by shane-tomlinson** [comment]
ref #2787 as the server side component.
